public/app/features/provisioning/Wizard/BootstrapStep.tsx

```diff
@@ -0,0 +1,398 @@
+          isDisabled: true,
+          disabledReason:
+            'Cannot pull to instance when you have existing resources. Please migrate your existing resources first.',
+        };
+      }
+
+      // Disable migrate option if there are existing files
+      if (option.operation === 'migrate' && fileCount > 0) {
+        return {
+          isDisabled: true,
+          disabledReason: 'Cannot migrate when repository has existing files. Please delete existing files first.',
+        };
+      }
+
+      if (option.value === 'instance' && otherInstanceConnected) {
+        return {
+          isDisabled: true,
+          disabledReason:
+            'Instance-wide connection is disabled because another instance is connected to this repository.',
+        };
+      }
+
+      if (option.value === 'instance' && otherFolderConnected) {
+        return {
+          isDisabled: true,
+          disabledReason: 'Instance-wide connection is disabled because folders are connected to repositories.',
+        };
+      }
+
+      return { isDisabled: false };
+    },
+    [
+      settingsQuery.data?.legacyStorage,
+      dashboardCount,
+      folderCount,
+      fileCount,
+      otherInstanceConnected,
+      otherFolderConnected,
+    ]
+  );
+
+  // Add sorted options after getOptionState is defined
+  const sortedModeOptions = useMemo(() => {
+    return [...modeOptions].sort((a, b) => {
+      const stateA = getOptionState(a);
+      const stateB = getOptionState(b);
+      return stateA.isDisabled === stateB.isDisabled ? 0 : stateA.isDisabled ? 1 : -1;
+    });
+  }, [getOptionState]);
+
+  const handleOptionSelect = useCallback(
+    (option: ModeOption) => {
+      const optionState = getOptionState(option);
+      if (optionState.isDisabled) {
+        return;
+      }
+
+      // Select the new option and update form state
+      setSelectedOption(option);
+      setValue('repository.sync.target', option.value);
+      onOptionSelect(option.operation === 'migrate');
+    },
+    [setValue, onOptionSelect, getOptionState]
+  );
+
+  const isLoading = settingsQuery.isLoading || isLoadingCounts || isLoadingFiles;
+  const hasAllData = Boolean(settingsQuery.data) && hasLoadedCounts && (!currentRepoName || filesData !== undefined);
+
+  // Store initialization function in a ref to avoid dependency issues
+  const initializeRef = useRef((options: typeof sortedModeOptions) => {
+    const firstAvailableOption = options.find((option) => !getOptionState(option).isDisabled);
+    if (firstAvailableOption) {
+      // Set form state first
+      setValue('repository.sync.target', firstAvailableOption.value);
+
+      // Small delay to ensure form state is set
+      setTimeout(() => {
+        setSelectedOption(firstAvailableOption);
+        onOptionSelect(firstAvailableOption.operation === 'migrate');
+        setHasInitialized(true);
+      }, 0);
+    } else {
+      setHasInitialized(true); // Mark as initialized even if no options available
```

Comment:
There'd be no need for any initialised state flags. Maybe moving all the options building logic int oa separate component will work?

Reply:
I tried to use the useAsync just now but it's calling the sync apis in a loop again which banned me on Github for the next hour.

Reply:
Make sure to have an empty dependency array for useAsync, so it's called only once.